xwiki-platform-core/xwiki-platform-display/xwiki-platform-display-api/src/main/java/org/xwiki/display/internal/DocumentReferenceDequeContext.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.display.internal;
+
+import java.util.Deque;
+import java.util.LinkedList;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.context.Execution;
+import org.xwiki.context.ExecutionContext;
+import org.xwiki.model.reference.DocumentReference;
+
+/**
+ * Provides a {@link Deque} that is stored in the execution context to store the stack of executed documents to avoid
+ * infinite recursion.
+ *
+ * @version $Id$
+ * @since 17.6.0RC1
+ * @since 17.4.3
+ * @since 16.10.10
+ */
+@Component(roles = DocumentReferenceDequeContext.class)
+@Singleton
+public class DocumentReferenceDequeContext
+{
+    /**
+     * The key used to store on the XWiki context map the stack of references to documents that are currently
+     * being evaluated (in the current execution context). This stack is used to prevent infinite recursion, which can
+     * happen if the title displayer is called on the current document from the title field or from a script within the
+     * first content heading.
+     */
+    private static final String DOCUMENT_REFERENCE_STACK_KEY = "internal.displayer.%s.documentReferenceStack";
+
+    /**
+     * Execution context handler, needed for accessing the XWiki context map.
+     */
+    @Inject
+    private Execution execution;
```

Comment:
I'm wondering if we should not go one step further in the bulletproofing and use directly a ThreadLocal here. I mean, it's always possible that a code in the middle of the stack start a fresh `ExecutionContext`. But that means we need some cleanup of that ThreadLocal in various places.

Reply:
Yes, true. I guess a cleanup at the start of the request might even be better. To me, it still seems like the execution context is what we want here and that we should be careful that we don't make it easy (for users) to trigger a completely new execution.